vm/src/stdlib/pystruct.rs

```diff
@@ -346,101 +505,125 @@ pub(crate) mod _struct {
-    ($T:ty) => {
-        paste::item! {
-            fn [<pack_ $T>]<Endianness>(vm: &VirtualMachine, arg: &PyObjectRef, data: &mut dyn Write) -> PyResult<()>
-            where
-                Endianness: byteorder::ByteOrder,
-            {
-                data.[<write_$T>]::<Endianness>(get_int_or_index(vm, arg)?).unwrap();
-                Ok(())
+                    fn unpack<Endianness: ByteOrder>(vm: &VirtualMachine, rdr: &mut dyn Read) -> PyResult {
+                        _unpack(vm, rdr, |rdr| rdr.[<read_$T>](), |i| Ok(i.into_pyobject(vm)))
+                    }
+                }
             }
-        }
-    };
-}
+        };
+    }
 
     macro_rules! make_pack_with_endianess {
-    ($T:ty) => {
-        paste::item! {
-            fn [<pack_ $T>]<Endianness>(vm: &VirtualMachine, arg: &PyObjectRef, data: &mut dyn Write) -> PyResult<()>
-            where
-                Endianness: byteorder::ByteOrder,
-            {
-                let v = $T::try_from_object(vm, arg.clone())?;
-                data.[<write_$T>]::<Endianness>(v).unwrap();
-                Ok(())
+        ($T:ty, $fromobj:path) => {
+            paste::item! {
+                impl Packable for $T {
+                    fn pack<Endianness: ByteOrder>(vm: &VirtualMachine, arg: &PyObjectRef, data: &mut dyn Write) -> PyResult<()> {
+                        data.[<write_$T>]::<Endianness>($fromobj(vm, arg.clone())?).unwrap();
+                        Ok(())
+                    }
+
+                    fn unpack<Endianness: ByteOrder>(vm: &VirtualMachine, rdr: &mut dyn Read) -> PyResult {
+                        _unpack(vm, rdr, |rdr| rdr.[<read_$T>]::<Endianness>(), |i| Ok(i.into_pyobject(vm)))
+                    }
+                }
             }
-        }
-    };
-}
+        };
+    }
 
     make_pack_no_endianess!(i8);
     make_pack_no_endianess!(u8);
-    make_pack_with_endianess_int!(i16);
-    make_pack_with_endianess_int!(u16);
-    make_pack_with_endianess_int!(i32);
-    make_pack_with_endianess_int!(u32);
-    make_pack_with_endianess_int!(i64);
-    make_pack_with_endianess_int!(u64);
-    make_pack_with_endianess!(f32);
-    make_pack_with_endianess!(f64);
-
-    fn pack_bool(vm: &VirtualMachine, arg: &PyObjectRef, data: &mut dyn Write) -> PyResult<()> {
-        let v = if IntoPyBool::try_from_object(vm, arg.clone())?.to_bool() {
-            1
-        } else {
-            0
-        };
-        data.write_u8(v).unwrap();
-        Ok(())
-    }
+    make_pack_with_endianess!(i16, get_int_or_index);
+    make_pack_with_endianess!(u16, get_int_or_index);
+    make_pack_with_endianess!(i32, get_int_or_index);
+    make_pack_with_endianess!(u32, get_int_or_index);
+    make_pack_with_endianess!(i64, get_int_or_index);
+    make_pack_with_endianess!(u64, get_int_or_index);
+    make_pack_with_endianess!(f32, TryFromObject::try_from_object);
+    make_pack_with_endianess!(f64, TryFromObject::try_from_object);
+
```

Comment:
CPython allows pack float format with int, float or class with __float__() but not a float like string. Non of our function doing exactly the job. We may need one it also can be use by array.array

Reply:
I think `IntoPyFloat::try_from_object` is correct; it accepts float, int, or anything with a `__float__` method

Reply:
I modified it in #2309 to not accept a float string, since that's what `PyFloat_AsDouble` does